packages/xdl/src/ProjectSettings.ts

```diff
@@ -135,6 +150,63 @@ export async function setPackagerInfoAsync(
   }
 }
 
+let devicesInfo: DevicesInfo | undefined;
+
+export async function readDevicesInfoAsync(projectRoot: string): Promise<DevicesInfo> {
+  if (devicesInfo) {
+    return devicesInfo;
+  }
+
+  try {
+    return await devicesJsonFile(projectRoot).readAsync({
+      cantReadFileDefault: { devices: [] },
+    });
+  } catch {
+    return await devicesJsonFile(projectRoot).writeAsync({ devices: [] });
+  }
+}
+
+export async function setDevicesInfoAsync(
+  projectRoot: string,
+  json: DevicesInfo
+): Promise<DevicesInfo> {
+  devicesInfo = json;
+
+  try {
+    return await devicesJsonFile(projectRoot).mergeAsync(json, {
+      cantReadFileDefault: { devices: [] },
+    });
+  } catch {
+    return await devicesJsonFile(projectRoot).writeAsync(json);
+  }
+}
+
+export async function saveDevicesAsync(
+  projectRoot: string,
+  deviceIds: string | string[]
+): Promise<void> {
+  const currentTime = new Date().getTime();
+  const newDeviceIds = typeof deviceIds === 'string' ? [deviceIds] : deviceIds;
+
+  const { devices } = await readDevicesInfoAsync(projectRoot);
+  const newDevicesJson = devices
+    .filter(device => {
+      if (newDeviceIds.includes(device.installationId)) {
+        return false;
+      }
+      // delete any devices that haven't been used to open this project in 30 days
```

Comment:
not sure intentional but it looks like this would only run when a manifest is requested. i think what would happen here is if it has been a few months since you worked on your project and go back and run `expo start` then we'd start dev sessions for all old devices and then shortly afterwards delete them, when the first manifest is loaded

Reply:
That's a fair point. I don't think it's a huge deal to start sessions for old devices once in that case, but on the other hand it's not hard to filter out old devices in the `read` function as well, so I'll go ahead and make that change.